--- src/lib.rs	2020-05-23 11:25:17.461931793 +0200
+++ src/lib.rs	2020-05-23 11:26:03.650562155 +0200
@@ -619,6 +619,7 @@
     RedirectUrl, RefreshToken, RefreshTokenRequest, RequestTokenError, Scope,
     StandardErrorResponse, StandardTokenResponse, TokenResponse as OAuth2TokenResponse, TokenType,
     TokenUrl, ResourceOwnerUsername, ResourceOwnerPassword, PasswordTokenRequest
+    , basic::BasicErrorResponseType
 };
 
 ///
